src/api/workspaceManager.ts

```diff
@@ -369,202 +171,167 @@ export default class WorkspaceManager {
-        } else if (
-            fs.readFileSync(pathToRootFile, { encoding: "utf-8" }) !== WORKSPACE_ROOT_FILE_TEXT
+    private async _refreshActiveCourseWorkspace(): Promise<Result<void, Error>> {
+        const workspaceName = this._getActiveCourseWorkspace();
+        const workspaceFolders = vscode.workspace.workspaceFolders;
+        if (!workspaceName || !workspaceFolders) {
+            Logger.warn("Attempted refresh for a non-course workspace.");
+            return Ok.EMPTY;
+        }
+
+        const rootFolder = this._resources.workspaceRootFolder;
+        const openExercises = this._exercises
+            .filter((x) => x.status === ExerciseStatus.Open)
+            .sort((a, b) => a.exerciseSlug.localeCompare(b.exerciseSlug))
+            .map((x) => ({ uri: x.uri }));
+        const correctStructure = [{ uri: rootFolder }, ...openExercises];
+        if (
+            _.zip(correctStructure, workspaceFolders).every(
+                ([a, b]) => a?.uri.fsPath === b?.uri.fsPath,
+            )
         ) {
-            Logger.log(`Rewriting ${WORKSPACE_ROOT_FILE_TEXT} at ${pathToRootFile}`);
-            fs.writeFileSync(pathToRootFile, WORKSPACE_ROOT_FILE_TEXT, { encoding: "utf-8" });
+            Logger.debug("Workspace refresh was a no-op.");
+            return Ok.EMPTY;
+        }
+
+        Logger.log("Refreshing workspace structure in the current workspace");
+        if (workspaceFolders[0]?.name !== WORKSPACE_ROOT_FOLDER_NAME) {
+            Logger.warn("Fixing incorrect root folder. This may restart the extension.");
+        }
+
+        const deleteCount = workspaceFolders.length;
+        Logger.debug(`Replacing ${deleteCount} workspace folders with ${correctStructure.length}`);
+        const success = vscode.workspace.updateWorkspaceFolders(
+            0,
+            deleteCount,
+            ...correctStructure,
+        );
+        if (!success) {
+            Logger.error("Replace operation failed.");
+            Logger.debug("Failed with folders:", ...correctStructure);
+        }
+
+        return success ? Ok.EMPTY : Err(new Error("Failed to refresh active workspace."));
+    }
+
+    private _onDidChangeWorkspaceFolders(e: vscode.WorkspaceFoldersChangeEvent): void {
+        const activeCourse = this.activeCourse;
+
+        let incorrectFolderAdded = false;
+        e.added.forEach((added) => {
+            const exercise = this._exercises.find((x) => x.uri === added.uri);
+            if (!exercise) {
+                incorrectFolderAdded = true;
+            } else if (exercise.courseSlug === activeCourse) {
+                exercise.status = ExerciseStatus.Open;
+            }
+        });
+
+        e.removed.forEach((removed) => {
+            const exercise = this._exercises.find((x) => x.uri === removed.uri);
+            if (exercise) {
+                exercise.status = ExerciseStatus.Closed;
+            }
+        });
+
```

Comment:
Collapse would be great to have as earlier for less cluttered code-workspace file:
https://github.com/rage/tmc-vscode/blob/d572e0218fb974bcb73da1e8ddb8106a0e5424be/src/init/index.ts#L83